src/builder/QuarkBuilder.sol

```diff
@@ -332,6 +332,166 @@ contract QuarkBuilder {
+        if (payment.isToken && !Strings.stringEqIgnoreCase(payment.currency, sellTokenSymbol)) {
+            // Bridge over payment token if not enough
+            uint256 maxCostOnDstChain = PaymentInfo.findMaxCost(payment, swapIntent.chainId);
+            if (needsBridgedFunds(payment.currency, maxCostOnDstChain, swapIntent.chainId, chainAccountsList, payment))
+            {
+                (IQuarkWallet.QuarkOperation[] memory bridgeQuarkOperations, Actions.Action[] memory bridgeActions) =
+                Actions.constructBridgeOperations(
+                    Actions.BridgeOperationInfo({
+                        assetSymbol: payment.currency,
+                        amountNeededOnDst: maxCostOnDstChain,
+                        dstChainId: swapIntent.chainId,
+                        recipient: swapIntent.sender,
+                        blockTimestamp: swapIntent.blockTimestamp,
+                        // TODO: set
+                        useQuotecall: false
+                    }),
+                    chainAccountsList,
+                    payment
+                );
+
+                for (uint256 i = 0; i < bridgeQuarkOperations.length; ++i) {
+                    quarkOperations[actionIndex] = bridgeQuarkOperations[i];
+                    actions[actionIndex] = bridgeActions[i];
+                    actionIndex++;
+                }
+            }
+        }
+
+        // Then, swap `amount` of `assetSymbol` to `recipient`
+        (quarkOperations[actionIndex], actions[actionIndex]) = Actions.matchaSwap(
+            Actions.MatchaSwap({
+                chainAccountsList: chainAccountsList,
+                entryPoint: swapIntent.entryPoint,
+                swapData: swapIntent.swapData,
+                sellToken: swapIntent.sellToken,
+                sellTokenSymbol: sellTokenSymbol,
+                sellAmount: swapIntent.sellAmount,
+                buyToken: swapIntent.buyToken,
+                buyTokenSymbol: buyTokenSymbol,
+                expectedBuyAmount: swapIntent.expectedBuyAmount,
+                chainId: swapIntent.chainId,
+                sender: swapIntent.sender,
+                blockTimestamp: swapIntent.blockTimestamp
+            }),
+            payment,
+            // TODO: Set this
+            false
+        );
+        actionIndex++;
+
+        // TODO: Merge transactions on same chain into Multicall. Maybe do that separately at the end via a helper function.
+
+        // Truncate actions and quark operations
+        actions = Actions.truncate(actions, actionIndex);
+        quarkOperations = Actions.truncate(quarkOperations, actionIndex);
+
+        // Validate generated actions for affordability
+        if (payment.isToken) {
+            assertSufficientPaymentTokenBalances(actions, chainAccountsList, swapIntent.chainId);
+        }
+
+        // Construct EIP712 digests
+        EIP712Helper.EIP712Data memory eip712Data;
```

Comment:
We can share this code at some point, right? In fact, most of the end of this function could (technically, not sure about Solidity-wise) just be a common function call, right?

Reply:
Yeah, this logic is being used in multiple places now. We'll want to clean it up a bit in a separate PR.